spec/immutable-data-structures.html

```diff
@@ -563,6 +563,72 @@ <h1>Tuple.prototype.with ( _index_, _value_ )</h1>
+            <p>_callbackfn_ is called with three arguments: the value of the element, the index of the element, and the Tuple being traversed.</p>
+            <p>The return value of `group` is an object that does not inherit from _Object.prototype_.</p>
+          </emu-note>
+          <p>When the `group` method is called with one or two arguments, the following steps are taken:</p>
+          <emu-alg>
+            1. Let _O_ be ? ToObject(*this* value).
+            1. Let _len_ be ? LengthOfArrayLike(_O_).
+            1. If IsCallable(_callbackfn_) is *false*, throw a *TypeError* exception.
+            1. Let _k_ be 0.
+            1. Let _groups_ be a new empty List.
+            1. Repeat, while _k_ &lt; _len_
+              1. Let _Pk_ be ! ToString(𝔽(_k_)).
+              1. Let _kValue_ be ? Get(_O_, _Pk_).
+              1. Let _propertyKey_ be ? ToPropertyKey(? Call(_callbackfn_, _thisArg_, &laquo; _kValue_, 𝔽(_k_), _O_ &raquo;)).
+              1. Perform AddValueToKeyedGroup(_groups_, _propertyKey_, _kValue_).
+              1. Set _k_ to _k_ + 1.
+            1. Let _obj_ be OrdinaryObjectCreate(*null*).
+            1. For each Record { [[Key]], [[Elements]] } _g_ of _groups_, do
+              1. Let _elements_ be CreateArrayFromList(_g_.[[Elements]]).
+              1. Perform ! CreateDataPropertyOrThrow(_obj_, _g_.[[Key]], _elements_).
+            1. Return _obj_.
+          </emu-alg>
+          <emu-note>
+            <p>The `group` function is intentionally generic; it does not require that its *this* value be a Tuple object. Therefore it can be transferred to other kinds of objects for use as a method.</p>
+          </emu-note>
+        </emu-clause>
+
+        <emu-clause id="sec-tuple.prototype.grouptomap">
+          <h1>Tuple.prototype.groupToMap ( _callbackfn_ [ , _thisArg_ ] )</h1>
+          <emu-note>
+            <p>_callbackfn_ should be a function that accepts three arguments. `groupToMap` calls _callbackfn_ once for each element in the Tuple, in ascending order, and constructs a new Map of arrays. Each value returned by _callbackfn_ is used as a key in the Map, and the associated element is included in the array in the constructed Map according to this key.</p>
+            <p>If a _thisArg_ parameter is provided, it will be used as the *this* value for each invocation of _callbackfn_. If it is not provided, *undefined* is used instead.</p>
+            <p>_callbackfn_ is called with three arguments: the value of the element, the index of the element, and the Tuple being traversed.</p>
+            <p>The return value of `groupToMap` is a Map.</p>
+          </emu-note>
+          <p>When the `groupToMap` method is called with one or two arguments, the following steps are taken:</p>
+          <emu-alg>
+            1. Let _O_ be ? ToObject(*this* value).
```

Comment:
The other tuple methods we defined with explicit steps all use `thisTupleValue(*this* value)` instead of `ToObject`.

Reply:
makes sense, I will adjust. I wasn't sure how much to deviate from the original Array versions of these methods.

Reply:
Do you think I should instead use the same methodology behind other Array methods like .at and just specify them in terms of the original? Or manually specify the algo here as well.

Reply:
manually is probably the better end goal regardless of how you do it in the proposal repo.